test/metabase/test/util/async.clj

```diff
@@ -30,3 +36,35 @@
          val))
      (finally
        (a/close! chan)))))
+
+(defn do-with-grouper-immediately-realizes!
+  [f]
+  (let [original-submit grouper/submit!
+        submitted       (atom [])
+        groupers        (atom #{})]
+    (with-redefs [grouper/submit! (fn [& args]
+                                    (swap! groupers conj (first args))
+                                    (swap! submitted conj
+                                           ;; this returns a promise we can wait for
+                                           (apply original-submit args)))]
+      (f (fn []
+           (u/poll {:thunk (fn []
+                             (doseq [grouper @groupers]
+                               (.wakeUp ^Grouper grouper))
+                             (every? realized? @submitted))
+                    :done? true?
+                    :interval 100
+                    :timeout 1000}))))))
+
+(defmacro with-grouper-realize!
```

Comment:
The term `realize` trips me up  bit here. Since we do not want any of our code to make use of the returned promises (see Slack thread) I suggest naming this something like `with-grouper-batches` and calling the binding `process-batch!`. Note the exclamation mark as the thunk has side-effects.

Reply:
renamed 83cdcb6775